lib/trino-filesystem/src/test/java/io/trino/filesystem/AbstractTestTrinoFileSystem.java

```diff
@@ -926,7 +930,12 @@ public void testFileWithTrailingWhitespace()
             }
             try (TrinoInputStream inputStream = inputFile.newStream()) {
                 byte[] bytes = ByteStreams.toByteArray(inputStream);
-                assertThat(bytes).isEqualTo(newContents);
+                if (!isFileContentCaching()) {
+                    assertThat(bytes).isEqualTo(newContents);
+                }
+                else {
+                    assertThat(bytes).isEqualTo(("test blob content for " + location).getBytes(UTF_8));
```

Comment:
Why aren't we getting `newContents` after write here ?
I think we should avoid having `isFileContentCaching` in this file

Reply:
The issue is that `lastModifiedTime` is cached on the `MemoryInputFile` (and on most `TrinoInputFile` implementations), so we get the same cache entry.

We could use a different `CacheKeyProvider`, but we also rely on `TrinoInputFile.length` being accurate for the files we read. Potentially we could skip the test entirely?

Reply:
The issue is:
1. We create `TrinoInputFile inputFile`.
2. We create a `TrinoInputStream` from the `inputFile` and read from it. For a caching file system this might use the file length on the `inputFile`. The file length is usually cached in the `inputFile`. The caching file system cannot currently control if the length is cached or not.
3. We overwrite the content of the input file. The cached file length of `inputFile` is now invalid.
4. We use the same `inputFile` to create a `TrinoInputStream`. This will use the cached length.
5. When we try to read using the `TrinoInputStream` we are using the wrong file length.

This is already an issue for the `HdfsInput.readTail`, which also relies on `TrinoInputFile.length`. If the test is modified to test `readTail` as below, it will fail for the HDFS and cache file systems, and succeed for the other file systems:
```java
    @Test
    public void testFileWithTrailingWhitespace()
            throws IOException
    {
        try (Closer closer = Closer.create()) {
            Location location = createBlob(closer, "dir/whitespace ");

            // Verify listing
            assertThat(listPath("dir")).isEqualTo(List.of(location));

            // Verify reading
            TrinoInputFile inputFile = getFileSystem().newInputFile(location);
            assertThat(inputFile.exists()).as("exists").isTrue();
            try (TrinoInputStream inputStream = inputFile.newStream()) {
                byte[] bytes = ByteStreams.toByteArray(inputStream);
                assertThat(bytes).isEqualTo(("test blob content for " + location).getBytes(UTF_8));
            }
            byte[] oldContents = ("test blob content for " + location).getBytes(UTF_8);
            // Read last 10 bytes
            try (TrinoInput input = inputFile.newInput()) {
                byte[] last10Bytes = new byte[10];
                input.readTail(last10Bytes, 0, 10);
                assertThat(last10Bytes).isEqualTo(Arrays.copyOfRange(oldContents, oldContents.length - 10, oldContents.length));
            }

            // Verify writing
            byte[] newContents = "bar bar baz new content aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa".getBytes(UTF_8);
            try (OutputStream outputStream = getFileSystem().newOutputFile(location).createOrOverwrite()) {
                outputStream.write(newContents.clone());
            }
            try (TrinoInputStream inputStream = inputFile.newStream()) {
                byte[] bytes = ByteStreams.toByteArray(inputStream);
                if (!isFileContentCaching()) {
                    assertThat(bytes).isEqualTo(newContents);
                }
                else {
                    assertThat(bytes).isEqualTo(("test blob content for " + location).getBytes(UTF_8));
                }
            }
            // Read last 10 bytes
            try (TrinoInput input = inputFile.newInput()) {
                byte[] last10Bytes = new byte[10];
                input.readTail(last10Bytes, 0, 10);
                assertThat(last10Bytes).isEqualTo(Arrays.copyOfRange(newContents, newContents.length - 10, newContents.length));
            }

            // Verify deleting
            getFileSystem().deleteFile(location);
            assertThat(inputFile.exists()).as("exists after delete").isFalse();

            // Verify renames
            if (supportsRenameFile()) {
                Location source = createBlob(closer, "dir/another trailing whitespace ");
                Location target = getRootLocation().appendPath("dir/after rename still whitespace ");
                getFileSystem().renameFile(source, target);
                assertThat(getFileSystem().newInputFile(source).exists()).as("source exists after rename").isFalse();
                assertThat(getFileSystem().newInputFile(target).exists()).as("target exists after rename").isTrue();

                try (TrinoInputStream inputStream = getFileSystem().newInputFile(target).newStream()) {
                    byte[] bytes = ByteStreams.toByteArray(inputStream);
                    assertThat(bytes).isEqualTo(("test blob content for " + source).getBytes(UTF_8));
                }

                getFileSystem().deleteFile(target);
                assertThat(getFileSystem().newInputFile(target).exists()).as("target exists after delete").isFalse();
            }
        }
    }
````

To leave the test unchanged we would need to either:
* Support invalidating the cached file length of each `TrinoInputFile` implementation.
* Not use the file length from the `TrinoInputFile` in any `TrinoFileSystemCache`.

Another option could be to change the test to create a new `TrinoInputFile` after the contents are modified.

Reply:
I think we can change the test to create a new TrinoInputFile after the contents are modified (with a code comment explaining why). In actual usage, typically files will not be overwritten. When they are overwritten, in flight reads might fail even without caching. But we should guarantee that after a file overwrite, a read using new TrinoInputFile succeeds and does not give stale results.

Reply:
ping